File: clients/python/src/termwright/framing.py
"""Wire framing: 4-byte big-endian unsigned length prefix + UTF-8 JSON body.

The declared length is checked against ``max_frame_bytes`` before any decoding.
Anything oversized, partial past the ceiling, or structurally hostile fails
closed with :class:`ProtocolViolation`, and a failed decoder never resumes:
resynchronising on an attacker-chosen offset is worse than dropping the link.
"""

from __future__ import annotations

import json
import re
from typing import Any, List, Optional

from .errors import ProtocolViolation

FRAME_HEADER_BYTES = 4

#: Property names that carry meaning in JavaScript engines. The reference
#: implementation rejects them, so we reject them too — a Python adapter must
#: not be able to smuggle a payload past a JS driver's projection.
RESERVED_KEYS = frozenset({"__proto__", "constructor", "prototype"})

_LONE_SURROGATE = re.compile("[\ud800-\udfff]")

_JSON_SEPARATORS = (",", ":")


def _reject_constant(name: str) -> Any:
    raise ProtocolViolation("dto-scalar", f"non-finite number literal {name}")


def _check_string(value: str, path: str, what: str) -> None:
    if _LONE_SURROGATE.search(value):
        raise ProtocolViolation("dto-string", f"unpaired surrogate in {what} at {path}")


def project_dto(value: Any, max_depth: int, _depth: int = 0, _path: str = "$") -> Any:
    """Deep-check an untrusted parsed value and return a JSON-safe copy.

    :param value: Untrusted input, typically the result of :func:`json.loads`.
    :param max_depth: Maximum nesting depth; the root sits at depth 0.
    :raises ProtocolViolation: On reserved keys, unpaired surrogates, non-finite
        numbers, non-JSON types, or nesting beyond ``max_depth``.
    """
    if _depth == 0 and (not isinstance(max_depth, int) or max_depth < 0):
        raise ProtocolViolation("dto-depth", "max_depth must be a non-negative integer")

    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        if value != value or value in (float("inf"), float("-inf")):
            raise ProtocolViolation("dto-scalar", f"non-finite number at {_path}")
        return value
    if isinstance(value, str):
        _check_string(value, _path, "string")
        return value

    if _depth > max_depth:
        raise ProtocolViolation("dto-depth", f"nesting exceeds {max_depth} at {_path}")

    if isinstance(value, list):
        return [
            project_dto(item, max_depth, _depth + 1, f"{_path}[{index}]")
            for index, item in enumerate(value)
        ]
    if isinstance(value, dict):
        projected = {}
        for key, item in value.items():
            if not isinstance(key, str):
                raise ProtocolViolation("dto-key", f"non-string key at {_path}")
            if key in RESERVED_KEYS:
                raise ProtocolViolation("dto-key", f'reserved property name "{key}" at {_path}')
            _check_string(key, _path, "key")
            projected[key] = project_dto(item, max_depth, _depth + 1, f"{_path}.{key}")
        return projected

    raise ProtocolViolation(
        "dto-scalar", f"value of type {type(value).__name__} is not JSON-representable at {_path}"
    )


def encode_json(message: Any) -> bytes:
    """Serialise to the canonical UTF-8 JSON body (no spaces, no NaN)."""
    try:
        text = json.dumps(message, ensure_ascii=False, separators=_JSON_SEPARATORS, allow_nan=False)
    except (TypeError, ValueError) as error:
        raise ProtocolViolation("frame-malformed", "message is not JSON-serialisable") from error
    try:
        return text.encode("utf-8")
    except UnicodeEncodeError as error:
        raise ProtocolViolation("dto-string", "message contains an unpaired surrogate") from error


def encode_frame(message: Any, max_frame_bytes: int) -> bytes:
    """Serialise a message into a single length-prefixed frame.

    :raises ProtocolViolation: If the value is not JSON-representable or the
        encoded body exceeds ``max_frame_bytes``.
    """
    _assert_ceiling(max_frame_bytes)
    body = encode_json(message)
    if len(body) > max_frame_bytes:
        raise ProtocolViolation(
            "frame-oversized",
            f"encoded frame is {len(body)} bytes, ceiling is {max_frame_bytes}",
        )
    return len(body).to_bytes(FRAME_HEADER_BYTES, "big") + body


def _assert_ceiling(max_frame_bytes: int) -> None:
    if not isinstance(max_frame_bytes, int) or isinstance(max_frame_bytes, bool) or max_frame_bytes <= 0:
        raise ProtocolViolation("frame-malformed", "max_frame_bytes must be a positive integer")


def decode_body(body: bytes, max_depth: int) -> Any:
    """Decode one frame body into a projected DTO."""
    try:
        text = body.decode("utf-8")
    except UnicodeDecodeError as error:
        raise ProtocolViolation("frame-encoding", "frame body is not valid UTF-8") from error
    try:
        parsed = json.loads(text, parse_constant=_reject_constant)
    except ProtocolViolation:
        raise
    except (ValueError, RecursionError) as error:
        raise ProtocolViolation("frame-malformed", "frame body is not valid JSON") from error
    return project_dto(parsed, max_depth)


class FrameDecoder:
    """Streaming decoder for length-prefixed JSON frames.

    Feed arbitrary chunks to :meth:`push`; it returns the messages that became
    complete. The first violation poisons the instance permanently.
    """

    def __init__(self, max_frame_bytes: int, max_depth: int) -> None:
        _assert_ceiling(max_frame_bytes)
        self._max_frame_bytes = max_frame_bytes
        self._max_depth = max_depth
        self._buffer = bytearray()
        self._failure: Optional[ProtocolViolation] = None

    @property
    def buffered(self) -> int:
        """Bytes held back waiting for the rest of a frame."""
        return len(self._buffer)

    def push(self, chunk: bytes) -> List[Any]:
        """Feed raw bytes; return the frames that completed, in order."""
        if self._failure is not None:
            raise ProtocolViolation(
                "decoder-poisoned",
                f"decoder failed earlier ({self._failure.code}) and accepts no further input",
            )
        try:
            return self._push_or_raise(chunk)
        except ProtocolViolation as error:
            self._failure = error
            self._buffer = bytearray()
            raise
        except Exception as error:  # pragma: no cover - defensive
            self._failure = ProtocolViolation("frame-malformed", "frame decoding failed")
            self._buffer = bytearray()
            raise self._failure from error

    def _push_or_raise(self, chunk: bytes) -> List[Any]:
        self._buffer.extend(chunk)
        messages: List[Any] = []
        offset = 0

        while len(self._buffer) - offset >= FRAME_HEADER_BYTES:
            length = int.from_bytes(self._buffer[offset : offset + FRAME_HEADER_BYTES], "big")
            if length == 0:
                raise ProtocolViolation("frame-malformed", "frame length must be non-zero")
            if length > self._max_frame_bytes:
                raise ProtocolViolation(
                    "frame-oversized",
                    f"frame declares {length} bytes, ceiling is {self._max_frame_bytes}",
                )
            end = offset + FRAME_HEADER_BYTES + length
            if len(self._buffer) < end:
                break
            messages.append(decode_body(bytes(self._buffer[offset + FRAME_HEADER_BYTES : end]), self._max_depth))
            offset = end

        if offset:
            del self._buffer[:offset]
        if len(self._buffer) > self._max_frame_bytes + FRAME_HEADER_BYTES:
            raise ProtocolViolation(
                "frame-oversized", f"buffered {len(self._buffer)} bytes without a complete frame"
            )
        return messages

File: clients/rust/src/framing.rs
//! Wire framing: a 4-byte big-endian length prefix and a UTF-8 JSON body.
//!
//! The declared length is checked against the ceiling before any body is
//! decoded, and the first violation poisons the decoder for good:
//! resynchronising on an attacker-chosen offset is worse than dropping the
//! connection.

use serde::Serialize;
use serde_json::Value;

use crate::error::Violation;

/// Size of the length prefix that precedes every frame body.
pub const FRAME_HEADER_BYTES: usize = 4;

/// Property names that carry meaning in JavaScript engines. The reference
/// implementation rejects them, so a Rust adapter cannot smuggle a payload
/// past a JS driver's projection either.
const RESERVED_KEYS: [&str; 3] = ["__proto__", "constructor", "prototype"];

/// Serialise a value into one length-prefixed frame.
///
/// Pass a [`serde_json::value::RawValue`] when the exact body bytes matter;
/// anything else is encoded by serde.
///
/// # Errors
/// Returns a [`Violation`] when the value is not JSON-encodable or the encoded
/// body exceeds `max_frame_bytes`.
pub fn encode_frame<T: Serialize>(value: &T, max_frame_bytes: usize) -> Result<Vec<u8>, Violation> {
    if max_frame_bytes == 0 {
        return Err(Violation::new(
            "frame-malformed",
            "maxFrameBytes must be positive",
        ));
    }
    let body = serde_json::to_vec(value)
        .map_err(|_| Violation::new("frame-malformed", "message is not JSON-serialisable"))?;
    if body.len() > max_frame_bytes {
        return Err(Violation::new(
            "frame-oversized",
            format!(
                "encoded frame is {} bytes, ceiling is {max_frame_bytes}",
                body.len()
            ),
        ));
    }
    let mut frame = Vec::with_capacity(FRAME_HEADER_BYTES + body.len());
    frame.extend_from_slice(&(body.len() as u32).to_be_bytes());
    frame.extend_from_slice(&body);
    Ok(frame)
}

/// Decode and check one frame body.
///
/// # Errors
/// Returns a [`Violation`] for non-UTF-8 bytes, invalid JSON, reserved
/// property names, or nesting beyond `max_depth`.
pub fn decode_body(body: &[u8], max_depth: usize) -> Result<Value, Violation> {
    let text = std::str::from_utf8(body)
        .map_err(|_| Violation::new("frame-encoding", "frame body is not valid UTF-8"))?;
    let value: Value = serde_json::from_str(text)
        .map_err(|_| Violation::new("frame-malformed", "frame body is not valid JSON"))?;
    project_dto(&value, max_depth)?;
    Ok(value)
}

/// Check an untrusted parsed value against the DTO rules.
///
/// Rust cannot express the getter, proxy or sparse-array cases the reference
/// implementation guards against — `serde_json` cannot produce them — so this
/// enforces what remains: reserved property names and the depth ceiling.
/// Non-finite numbers cannot survive JSON parsing at all.
///
/// # Errors
/// Returns a [`Violation`] with code `dto-key` or `dto-depth`.
pub fn project_dto(value: &Value, max_depth: usize) -> Result<(), Violation> {
    project_node(value, 0, max_depth, "$")
}

fn project_node(
    value: &Value,
    depth: usize,
    max_depth: usize,
    path: &str,
) -> Result<(), Violation> {
    match value {
        Value::Array(items) => {
            if depth > max_depth {
                return Err(depth_violation(max_depth, path));
            }
            for (index, item) in items.iter().enumerate() {
                project_node(item, depth + 1, max_depth, &format!("{path}[{index}]"))?;
            }
            Ok(())
        }
        Value::Object(entries) => {
            if depth > max_depth {
                return Err(depth_violation(max_depth, path));
            }
            for (key, item) in entries {
                if RESERVED_KEYS.contains(&key.as_str()) {
                    return Err(Violation::new(
                        "dto-key",
                        format!("reserved property name \"{key}\" at {path}"),
                    ));
                }
                project_node(item, depth + 1, max_depth, &format!("{path}.{key}"))?;
            }
            Ok(())
        }
        _ => Ok(()),
    }
}

fn depth_violation(max_depth: usize, path: &str) -> Violation {
    Violation::new(
        "dto-depth",
        format!("nesting exceeds {max_depth} at {path}"),
    )
}

/// One decoded wire frame: the raw body plus its parsed value.
#[derive(Debug, Clone)]
pub struct Frame {
    /// The body bytes exactly as they arrived.
    pub raw: Vec<u8>,
    /// The parsed, checked value.
    pub value: Value,
}

/// Streaming decoder for length-prefixed JSON frames.
#[derive(Debug)]
pub struct FrameDecoder {
    max_frame_bytes: usize,
    max_depth: usize,
    buffer: Vec<u8>,
    failure: Option<Violation>,
}

impl FrameDecoder {
    /// Create a decoder bounded by `max_frame_bytes` and `max_depth`.
    pub fn new(max_frame_bytes: usize, max_depth: usize) -> Self {
        Self {
            max_frame_bytes,
            max_depth,
            buffer: Vec::new(),
            failure: None,
        }
    }

    /// Bytes held back waiting for the rest of a frame.
    pub fn buffered(&self) -> usize {
        self.buffer.len()
    }

    /// Feed raw bytes and take the frames that completed, in order.
    ///
    /// # Errors
    /// Any violation is returned once and then latched: later calls fail with
    /// `decoder-poisoned`.
    pub fn push(&mut self, chunk: &[u8]) -> Result<Vec<Frame>, Violation> {
        if let Some(failure) = &self.failure {
            return Err(Violation::new(
                "decoder-poisoned",
                format!(
                    "decoder failed earlier ({}) and accepts no further input",
                    failure.code
                ),
            ));
        }
        match self.push_inner(chunk) {
            Ok(frames) => Ok(frames),
            Err(violation) => {
                self.failure = Some(violation.clone());
                self.buffer = Vec::new();
                Err(violation)
            }
        }
    }

    fn push_inner(&mut self, chunk: &[u8]) -> Result<Vec<Frame>, Violation> {
        self.buffer.extend_from_slice(chunk);
        let mut frames = Vec::new();
        let mut offset = 0usize;

        while self.buffer.len() - offset >= FRAME_HEADER_BYTES {
            let header: [u8; FRAME_HEADER_BYTES] = self.buffer[offset..offset + FRAME_HEADER_BYTES]
                .try_into()
                .expect("slice length checked above");
            let length = u32::from_be_bytes(header) as usize;
            if length == 0 {
                return Err(Violation::new(
                    "frame-malformed",
                    "frame length must be non-zero",
                ));
            }
            if length > self.max_frame_bytes {
                return Err(Violation::new(
                    "frame-oversized",
                    format!(
                        "frame declares {length} bytes, ceiling is {}",
                        self.max_frame_bytes
                    ),
                ));
            }
            let end = offset + FRAME_HEADER_BYTES + length;
            if self.buffer.len() < end {
                break;
            }
            let body = &self.buffer[offset + FRAME_HEADER_BYTES..end];
            let value = decode_body(body, self.max_depth)?;
            frames.push(Frame {
                raw: body.to_vec(),
                value,
            });
            offset = end;
        }

        if offset > 0 {
            self.buffer.drain(..offset);
        }
        if self.buffer.len() > self.max_frame_bytes + FRAME_HEADER_BYTES {
            return Err(Violation::new(
                "frame-oversized",
                format!(
                    "buffered {} bytes without a complete frame",
                    self.buffer.len()
                ),
            ));
        }
        Ok(frames)
    }
}
